icu4c/source/common/wintz.cpp

```diff
@@ -47,77 +48,61 @@ U_NAMESPACE_BEGIN
 U_INTERNAL const char* U_EXPORT2
 uprv_detectWindowsTimeZone()
 {
-    UErrorCode status = U_ZERO_ERROR;
-    char* icuid = nullptr;
-    char dynamicTZKeyName[MAX_TIMEZONE_ID_LENGTH];
-    char tmpid[MAX_TIMEZONE_ID_LENGTH];
-    int32_t len;
-    int id = GEOID_NOT_AVAILABLE;
-    int errorCode;
-    wchar_t ISOcodeW[3] = {}; /* 2 letter ISO code in UTF-16 */
-    char ISOcode[3] = {}; /* 2 letter ISO code in UTF-8 */
-
+    /* Obtain TIME_ZONE_INFORMATION from the API and get the non-localized time zone name. */
     DYNAMIC_TIME_ZONE_INFORMATION dynamicTZI;
     uprv_memset(&dynamicTZI, 0, sizeof(dynamicTZI));
-    uprv_memset(dynamicTZKeyName, 0, sizeof(dynamicTZKeyName));
-    uprv_memset(tmpid, 0, sizeof(tmpid));
 
-    /* Obtain TIME_ZONE_INFORMATION from the API and get the non-localized time zone name. */
-    if (TIME_ZONE_ID_INVALID == GetDynamicTimeZoneInformation(&dynamicTZI)) {
+    DWORD tzIDStatus = GetDynamicTimeZoneInformation(&dynamicTZI);
+    if (tzIDStatus == TIME_ZONE_ID_INVALID || dynamicTZI.TimeZoneKeyName[0] == 0)
         return nullptr;
-    }
-
-    id = GetUserGeoID(GEOCLASS_NATION);
-    errorCode = GetGeoInfoW(id, GEO_ISO2, ISOcodeW, 3, 0);
 
-    // convert from wchar_t* (UTF-16 on Windows) to char* (UTF-8).
-    u_strToUTF8(ISOcode, UPRV_LENGTHOF(ISOcode), nullptr,
-        reinterpret_cast<const UChar*>(ISOcodeW), UPRV_LENGTHOF(ISOcodeW), &status);
+    // If DST is turned off in the control panel, return "Etc/GMT<offset>".
+    if (dynamicTZI.DynamicDaylightTimeDisabled) {
+        LONG utcOffsetMins = dynamicTZI.Bias;
+        if (utcOffsetMins == 0)
+          return uprv_strdup("UTC");
+
+        // No way to support when DST is turned off and the offset in minutes is not
+        // a multiple of 60.
+        if (utcOffsetMins % 60 == 0) {
+          char gmtOffsetTz[11]; // "Etc/GMT+dd" is 11-char long with a terminal null.
+          // Note '-' before 'utcOffsetMin'. The timezone ID's sign convention
+          // is that a timezone ahead of UTC is Etc/GMT-<offset> and a timezone
+          // behind UTC is Etc/GMT+<offset>.
+          snprintf(gmtOffsetTz, 11, "Etc/GMT%+d", -utcOffsetMins / 60);
```

Comment:
I think it might also be good to check the return from `snprintf`, just to make sure that we don't exceed the buffer size of 11.
If we do, then there was somehow a bad value obtained from the `dynamicTZI.Bias`. This shouldn't happen, but it might be better to be defensive perhaps.